datasource/amazon/ami/data.go

```diff
@@ -0,0 +1,28 @@
+package ami
+
+import (
+	"github.com/hashicorp/hcl/v2/hcldec"
+	"github.com/zclconf/go-cty/cty"
+)
+
+type DataSource struct {
+	config Config
+}
+
+func (d *DataSource) ConfigSpec() hcldec.ObjectSpec {
+	return d.config.FlatMapstructure().HCL2Spec()
+}
+
+func (d *DataSource) OutputSpec() hcldec.ObjectSpec {
+	return hcldec.ObjectSpec{
+		"amazon-ami": &hcldec.LiteralSpec{Value: cty.StringVal("placeholder")},
+	}
```

Comment:
A datasource can return nested hcl 'objects' and I think it's really important we work with objects here and I can help with that if you need ! 🙂 
The AMI datasource already should return an object that could look like:

```go
// DataSourceOutput represents the possible fields returned by this datasource.
type DataSourceOutput struct {
  ID           string
  Name         string
  CreationDate string
  Owner        string
  OwnerName    string
  Tags         map[string]string
}
```
One thing we can do here is :

* create a go struct that represents the datasource return type ( just like for a plugin )
* generate the hcl2 spec code for it.
* and use that func here:
```suggestion
	return DataSourceOutput{}.FlatMapstructure().HCL2Spec()
```
* This could also auto-gen some docs for us !!!

Reply:
On the same topic, I know that we've talked on zoom about the fact that unknown values and specs don't go through the wire using gob/rpc.

We could add some code to generate a 'mock' value for such a struct so we don't have to fix that, but I think that the time this will take should rather be spent on making this work with unknown values of the type defined by `OutputSpec`. Your call !

Reply:
I would very much enjoy learning more about go code. FRom what I've seen it is very useful and a huge asset.